src/passes/OptimizeInstructions.cpp

```diff
@@ -974,25 +950,36 @@ struct OptimizeInstructions
   Expression* optimizeBoolean(Expression* boolean) {
     // TODO use a general getFallthroughs
     if (auto* unary = boolean->dynCast<Unary>()) {
-      if (unary && unary->op == EqZInt32) {
-        auto* unary2 = unary->value->dynCast<Unary>();
-        if (unary2 && unary2->op == EqZInt32) {
-          // double eqz
-          return unary2->value;
+      if (unary) {
+        if (unary->op == EqZInt32 || unary->op == EqZInt64) {
+          if (unary->op == EqZInt32) {
+            auto* unary2 = unary->value->dynCast<Unary>();
+            if (unary2 && unary2->op == EqZInt32) {
+              // double eqz
+              return unary2->value;
```

Comment:
this seems separate from the select issue, and could be another PR? also the rest below.

Reply:
Currently this PR utilize improved `optimizeBoolean` so better update all together

Reply:
I do still think it would be faster to split these up. I won't block it on that. But as you can see there are often small bugs or other issues that I end up finding. The more changes in a single PR, the longer it takes for it to land...